components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/DefaultURLResolverService.java

```diff
@@ -0,0 +1,201 @@
+        }
+        StringBuilder serverUrl = new StringBuilder(mgtTransport).append("://").append(hostName.toLowerCase());
+        // If it's well known HTTPS port, skip adding port.
+        if (mgtTransportPort != IdentityCoreConstants.DEFAULT_HTTPS_PORT) {
+            serverUrl.append(":").append(mgtTransportPort);
+        }
+
+        appendContextToUri(urlContext, addProxyContextPath, addWebContextRoot, serverUrl, addTenantQueryParamInLegacyMode);
+
+        return serverUrl.toString();
+    }
+
+    private int getMgtTransportPort(String mgtTransport) {
+
+        AxisConfiguration axisConfiguration = IdentityCoreServiceComponent.getConfigurationContextService().
+                getServerConfigContext().getAxisConfiguration();
+        int mgtTransportPort = CarbonUtils.getTransportProxyPort(axisConfiguration, mgtTransport);
+        if (mgtTransportPort <= 0) {
+            mgtTransportPort = CarbonUtils.getTransportPort(axisConfiguration, mgtTransport);
+        }
+        return mgtTransportPort;
+    }
+
+    private String getHostName() throws URLResolverException {
+
+        String hostName = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants.HOST_NAME);
+        try {
+            if (hostName == null) {
+                hostName = NetworkUtils.getLocalHostname();
+            }
+        } catch (SocketException e) {
+            throw new URLResolverException("Error while trying to read hostname.", e);
+        }
+        return hostName;
+    }
+
+    private void appendWebContextRoot(StringBuilder serverUrl) {
+
+        String webContextRoot = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants
+                .WEB_CONTEXT_ROOT);
+        // If webContextRoot is defined then append it.
+        if (StringUtils.isNotBlank(webContextRoot)) {
+            if (webContextRoot.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(webContextRoot.trim());
+            } else {
+                serverUrl.append(webContextRoot.trim());
+            }
+        }
+    }
+
+    private void appendProxyContextPath(StringBuilder serverUrl) {
+
+        String proxyContextPath = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants
+                .PROXY_CONTEXT_PATH);
+        // If ProxyContextPath is defined then append it.
+        if (StringUtils.isNotBlank(proxyContextPath)) {
+            if (proxyContextPath.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(proxyContextPath.trim());
+            } else {
+                serverUrl.append(proxyContextPath.trim());
+            }
+        }
+    }
+
+    private void appendTenantAsPathParam(StringBuilder serverUrl) {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        }
```

Comment:
We will be having this empty check everywhere right. What if we move this to the same util.
@mefarazath WDYT?

Reply:
Thinking again I think this is the way we should proceed as there can be ways tenant will be available based on context